### PR TITLE
Fix for next/head issues

### LINKED ISSUE #17721: Failed to execute 'createElement' on 'Document' | v9.5.4
# Bug report

## Describe the bug

I ran into this error after updating Next.js to version 9.5.4. The problem is that you cannot use React Components inside `Head` in _app.js file.

**Minimal Example:**

```
const Metadata = () => (
  <>
      <base href="/" />
      <meta name="msapplication-TileColor" content="#000000" />
      {/*...*/}
  </>
)

class NextApp extends App {
  render() {
      return (
          <>
              <Head>
                  <title>Next Bug Report</title>
                  {/*Error*/}
                  <Metadata />
              </Head>
          </>
      )
  }
}
```

## To Reproduce

Just open the dev server

**Reproduction repo:** [next-bug-report](https://github.com/arthruwelkin/next-bug-report)

## Screenshots

image

## System information

- Version of Next.js: [9.5.4]
- Version of React/ReactDOM: [16.13.1]
- Version of Node.js: [14.0.0]

### LINKED ISSUE #17775: [Next 9.5.4 or above] Throw function expected error and app stops working in Edge 41
# Bug report

## Describe the bug
App throws SCRIPT5002: Function expected error in Edge 41 or Edge 42. 
Some react components are broken, all links and buttons are stop working.

## To Reproduce
Open a next.js 9.5.4 or above application in production mode on Browserstack with preset Edge 41 or Edge 42.

## Expected behavior
Next.js should transpile the script to ES5(supports Set) and running without crash in Edge 41 and Edge 42

## Screenshots
Error message
Screen Shot 2020-10-10 at 7 53 46 pm

Error highlighted in debugger
Screen Shot 2020-10-10 at 7 50 30 pm

## System information

- OS: macOS
- Browser: 
Edge 41.16299.15.0
Edge HTML 16.16299

Edge 42.17134.1098.0
Edge HTML 17.17134

- Version of Next.js: 9.5.4 or above
- Version of Node.js: 14.13.1

## Additional context
These new changes in head-manager.js/ts which break the app since Next 9.5.4
`const headEl = document.getElementsByTagName('head')[0]`
` const elements = new Set<Element>(headEl.children)`
https://github.com/vercel/next.js/blob/658810815035e55a7031f27c5a6f3c01baa31ccf/packages/next/client/head-manager.ts#L92
I have to rollback to Next 9.5.3 in order to get the app running in Edge 41 and 42.

### PR DESCRIPTION
**Pull request by devknoll**
Addresses two issues that came up when removing `next-head-count` in #16758. 

---

Fixes #17775
Fixes #17721

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| buildDuration | 11.6s | 11.7s | ⚠️ +142ms |
| nodeModulesSize | 64.3 MB | 64.3 MB | ⚠️ +565 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.148 | 2.164 | ⚠️ +0.02 |
| / avg req/sec | 1163.89 | 1155.29 | ⚠️ -8.6 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.194 | 1.172 | -0.02 |
| /error-in-render avg req/sec | 2093.74 | 2133.31 | +39.57 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| 677f882d2ed8..0989.js gzip | 11.1 kB | 11.1 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-43e96dd..7c44.js gzip | 7.3 kB | 7.31 kB | ⚠️ +5 B |
| webpack-e067..f178.js gzip | 751 B | 751 B | ✓ |
| Overall change | 58.1 kB | 58.1 kB | ⚠️ +5 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| 677f882d2ed8..dule.js gzip | 6.91 kB | 6.91 kB | ✓ |
| framework.HA..dule.js gzip | 39 kB | 39 kB | ✓ |
| main-ee66c17..dule.js gzip | 6.29 kB | 6.3 kB | ⚠️ +5 B |
| webpack-07c5..dule.js gzip | 751 B | 751 B | ✓ |
| Overall change | 52.9 kB | 52.9 kB | ⚠️ +5 B |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| polyfills-4b..e242.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| _app-9a0b9e1..b37e.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-ed1b0..8fbd.js gzip | 3.44 kB | 3.44 kB | ✓ |
| hooks-89731c..c609.js gzip | 887 B | 887 B | ✓ |
| index-17468f..5d83.js gzip | 227 B | 227 B | ✓ |
| link-89ad9e7..25bb.js gzip | 1.34 kB | 1.34 kB | ✓ |
| routerDirect..924c.js gzip | 284 B | 284 B | ✓ |
| withRouter-7..c13d.js gzip | 284 B | 284 B | ✓ |
| Overall change | 7.74 kB | 7.74 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| _app-75d3a82..dule.js gzip | 625 B | 625 B | ✓ |
| _error-4469a..dule.js gzip | 2.29 kB | 2.29 kB | ✓ |
| hooks-cbf13f..dule.js gzip | 387 B | 387 B | ✓ |
| index-b9a643..dule.js gzip | 226 B | 226 B | ✓ |
| link-aeb707b..dule.js gzip | 1.29 kB | 1.29 kB | ✓ |
| routerDirect..dule.js gzip | 284 B | 284 B | ✓ |
| withRouter-f..dule.js gzip | 282 B | 282 B | ✓ |
| Overall change | 5.39 kB | 5.39 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| _buildManifest.js gzip | 322 B | 322 B | ✓ |
| _buildManife..dule.js gzip | 329 B | 329 B | ✓ |
| Overall change | 651 B | 651 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| index.html gzip | 1 kB | 1 kB | ✓ |
| link.html gzip | 1.01 kB | 1.01 kB | ⚠️ +1 B |
| withRouter.html gzip | 995 B | 996 B | ⚠️ +1 B |
| Overall change | 3 kB | 3.01 kB | ⚠️ +2 B |

</details>

#### Diffs
<details>
<summary>Diff for <strong>main-32c48d3..47fa3ff41.js</strong></summary>

```diff
@@ -182,8 +182,10 @@ _N_E = (window["webpackJsonp_N_E"] = window["webpackJsonp_N_E"] || []).push([
       }
 
       function initHeadManager(initialHeadEntries) {
-        var headEl = document.getElementsByTagName("head")[0];
-        var elements = new Set(headEl.children);
+        var headEl = document.getElementsByTagName("head")[0]; // Some browsers like Edge 16 don't support constructing `Set` with an `HTMLCollection`,
+        // so we need to first convert it using `Array.from`.
+
+        var elements = new Set(Array.from(headEl.children));
         updateElements(
           elements,
           initialHeadEntries.map(function(_ref2) {
```
</details>
<details>
<summary>Diff for <strong>main-f1e531e..6f.module.js</strong></summary>

```diff
@@ -151,8 +151,10 @@
       }
 
       function initHeadManager(initialHeadEntries) {
-        var headEl = document.getElementsByTagName("head")[0];
-        var elements = new Set(headEl.children);
+        var headEl = document.getElementsByTagName("head")[0]; // Some browsers like Edge 16 don't support constructing `Set` with an `HTMLCollection`,
+        // so we need to first convert it using `Array.from`.
+
+        var elements = new Set(Array.from(headEl.children));
         updateElements(
           elements,
           initialHeadEntries.map(_ref2 => {
```
</details>
<details>
<summary>Diff for <strong>index.html</strong></summary>

```diff
@@ -6,7 +6,7 @@
     <noscript data-n-css="true"></noscript>
     <link
       rel="preload"
-      href="/_next/static/chunks/main-f1e531e41c50db5e466f.module.js"
+      href="/_next/static/chunks/main-c006abc06f02d320d334.module.js"
       as="script"
       crossorigin="anonymous"
     />
@@ -85,13 +85,13 @@
       src="/_next/static/chunks/polyfills-f73ba3fc145972ef83e9.js"
     ></script>
     <script
-      src="/_next/static/chunks/main-32c48d3fa8f47fa3ff41.js"
+      src="/_next/static/chunks/main-30f0b73c8660372816b8.js"
       async=""
       crossorigin="anonymous"
       nomodule=""
     ></script>
     <script
-      src="/_next/static/chunks/main-f1e531e41c50db5e466f.module.js"
+      src="/_next/static/chunks/main-c006abc06f02d320d334.module.js"
       async=""
       crossorigin="anonymous"
       type="module"
```
</details>
<details>
<summary>Diff for <strong>link.html</strong></summary>

```diff
@@ -6,7 +6,7 @@
     <noscript data-n-css="true"></noscript>
     <link
       rel="preload"
-      href="/_next/static/chunks/main-f1e531e41c50db5e466f.module.js"
+      href="/_next/static/chunks/main-c006abc06f02d320d334.module.js"
       as="script"
       crossorigin="anonymous"
     />
@@ -90,13 +90,13 @@
       src="/_next/static/chunks/polyfills-f73ba3fc145972ef83e9.js"
     ></script>
     <script
-      src="/_next/static/chunks/main-32c48d3fa8f47fa3ff41.js"
+      src="/_next/static/chunks/main-30f0b73c8660372816b8.js"
       async=""
       crossorigin="anonymous"
       nomodule=""
     ></script>
     <script
-      src="/_next/static/chunks/main-f1e531e41c50db5e466f.module.js"
+      src="/_next/static/chunks/main-c006abc06f02d320d334.module.js"
       async=""
       crossorigin="anonymous"
       type="module"
```
</details>
<details>
<summary>Diff for <strong>withRouter.html</strong></summary>

```diff
@@ -6,7 +6,7 @@
     <noscript data-n-css="true"></noscript>
     <link
       rel="preload"
-      href="/_next/static/chunks/main-f1e531e41c50db5e466f.module.js"
+      href="/_next/static/chunks/main-c006abc06f02d320d334.module.js"
       as="script"
       crossorigin="anonymous"
     />
@@ -85,13 +85,13 @@
       src="/_next/static/chunks/polyfills-f73ba3fc145972ef83e9.js"
     ></script>
     <script
-      src="/_next/static/chunks/main-32c48d3fa8f47fa3ff41.js"
+      src="/_next/static/chunks/main-30f0b73c8660372816b8.js"
       async=""
       crossorigin="anonymous"
       nomodule=""
     ></script>
     <script
-      src="/_next/static/chunks/main-f1e531e41c50db5e466f.module.js"
+      src="/_next/static/chunks/main-c006abc06f02d320d334.module.js"
       async=""
       crossorigin="anonymous"
       type="module"
```
</details>
</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| buildDuration | 13.7s | 13.4s | -308ms |
| nodeModulesSize | 64.3 MB | 64.3 MB | ⚠️ +565 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| 677f882d2ed8..0989.js gzip | 11.1 kB | 11.1 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-43e96dd..7c44.js gzip | 7.3 kB | N/A | N/A |
| webpack-e067..f178.js gzip | 751 B | 751 B | ✓ |
| main-f7ffcb2..6e46.js gzip | N/A | 7.31 kB | N/A |
| Overall change | 58.1 kB | 58.1 kB | ⚠️ +5 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| 677f882d2ed8..dule.js gzip | 6.91 kB | 6.91 kB | ✓ |
| framework.HA..dule.js gzip | 39 kB | 39 kB | ✓ |
| main-ee66c17..dule.js gzip | 6.29 kB | N/A | N/A |
| webpack-07c5..dule.js gzip | 751 B | 751 B | ✓ |
| main-367e4e6..dule.js gzip | N/A | 6.3 kB | N/A |
| Overall change | 52.9 kB | 52.9 kB | ⚠️ +5 B |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| polyfills-4b..e242.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| _app-9a0b9e1..b37e.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-ed1b0..8fbd.js gzip | 3.44 kB | 3.44 kB | ✓ |
| hooks-89731c..c609.js gzip | 887 B | 887 B | ✓ |
| index-17468f..5d83.js gzip | 227 B | 227 B | ✓ |
| link-89ad9e7..25bb.js gzip | 1.34 kB | 1.34 kB | ✓ |
| routerDirect..924c.js gzip | 284 B | 284 B | ✓ |
| withRouter-7..c13d.js gzip | 284 B | 284 B | ✓ |
| Overall change | 7.74 kB | 7.74 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| _app-75d3a82..dule.js gzip | 625 B | 625 B | ✓ |
| _error-4469a..dule.js gzip | 2.29 kB | 2.29 kB | ✓ |
| hooks-cbf13f..dule.js gzip | 387 B | 387 B | ✓ |
| index-b9a643..dule.js gzip | 226 B | 226 B | ✓ |
| link-aeb707b..dule.js gzip | 1.29 kB | 1.29 kB | ✓ |
| routerDirect..dule.js gzip | 284 B | 284 B | ✓ |
| withRouter-f..dule.js gzip | 282 B | 282 B | ✓ |
| Overall change | 5.39 kB | 5.39 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| _buildManifest.js gzip | 322 B | 322 B | ✓ |
| _buildManife..dule.js gzip | 329 B | 329 B | ✓ |
| Overall change | 651 B | 651 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | azukaru/next.js fix-next-head-components | Change |
| - | - | - | - |
| _error.js | 1.06 MB | 1.06 MB | ⚠️ +61 B |
| 404.html | 4.34 kB | 4.34 kB | ✓ |
| hooks.html | 3.92 kB | 3.92 kB | ✓ |
| index.js | 1.06 MB | 1.06 MB | ⚠️ +61 B |
| link.js | 1.1 MB | 1.1 MB | ⚠️ +61 B |
| routerDirect.js | 1.1 MB | 1.1 MB | ⚠️ +61 B |
| withRouter.js | 1.1 MB | 1.1 MB | ⚠️ +61 B |
| Overall change | 5.41 MB | 5.41 MB | ⚠️ +305 B |

</details>

</details>
Commit: 2e52c0c43c392aba8651964c96eb25d737ff4f6c

**Comment by jflayhart** [comment]
Nice! In case it helps, I added some tests around this in [a similar PR](https://github.com/vercel/next.js/pull/17770). Would be nice to collab since it seems you understand how all this head stuff works.

File: packages/next/client/head-manager.ts [hunk 1]
```diff
@@ -89,7 +89,9 @@ function updateElements(
 
 export default function initHeadManager(initialHeadEntries: HeadEntry[]) {
   const headEl = document.getElementsByTagName('head')[0]
-  const elements = new Set<Element>(headEl.children)
+  // Some browsers like Edge 16 don't support constructing `Set` with an `HTMLCollection`,
+  // so we need to first convert it using `Array.from`.
+  const elements = new Set<Element>(Array.from(headEl.children))
```
Comment (jflayhart) [inline packages/next/client/head-manager.ts:94]: For posterity, this won't fix: #17721 per @timneutkens comments on it not working for client-side navigation.

Comment (devknoll) [reply packages/next/client/head-manager.ts:94]: Could you elaborate on your comment, please? Are you just referring to custom components not being client-side rendered?

Comment (jflayhart) [reply packages/next/client/head-manager.ts:94]: Referring to https://github.com/vercel/next.js/issues/17721#issuecomment-709487069 which is the "previous" behavior in 9.5.3, which I think this PR will get us back to that, but not solve the issue. But you've explained further concerns about this in #17721.

**Review by styfle — APPROVED** [review]

**Pull request closed without merging**